### PR TITLE
[libc] default enable -ftrivial-auto-var-init=pattern

### PR DESCRIPTION
**Pull request by nickdesaulniers**
Usage of uninitialized memory is a top memory safety issue in C++ codebases.
Help mitigate this somewhat by default initialize stack allocations to a
pattern (0xAA repeating).

Clang has received optimizations to sink these into control flow paths that
access such values to minimize the overhead of these added initializations.

If there's a measurable slowdown, we can add
-ftrivial-auto-var-init-max-size=<N> for some value N bytes if we have any
large stack allocations, or add attribute uninitialized to any variable
declarations.

Unsupported until GCC 12.1 / Clang 8.

Increases file size of libc.a from a full build by +8.79Ki (+0.2%).

### CODE REVIEW

**Comment by nickdesaulniers** [comment]
FWIW, here's some measurements from bloaty: https://gist.github.com/nickdesaulniers/1092b3748b6810ae97053973c40b19a2.

The two cases of relatively large increases in code size:
1. `__llvm_libc_18_0_0_git::StrToNumResult<__llvm_libc_18_0_0_git::internal::ExpandedFloat<long double> > __llvm_libc_18_0_0_git::internal::hexadecimal_string_to_float<long double>(char const*, char, __llvm_libc_18_0_0_git::internal::RoundDirection)`
2. `__llvm_libc_18_0_0_git::internal::FloatConvertReturn<long double> __llvm_libc_18_0_0_git::internal::decimal_exp_to_float<long double>(__llvm_libc_18_0_0_git::internal::ExpandedFloat<long double>, char const*, bool, __llvm_libc_18_0_0_git::internal::RoundDirection)`

These look like their no longer being inlined due to increases in code size, perhaps:
1. `__llvm_libc_18_0_0_git::construct_thread_name_file_path(__llvm_libc_18_0_0_git::cpp::StringStream&, int)`
2. `__llvm_libc_18_0_0_git::printf_core::FloatWriter::write_middle_block(unsigned int)`

File: libc/cmake/modules/LLVMLibCObjectRules.cmake [hunk 1]
```diff
@@ -41,6 +41,11 @@ function(_get_common_compile_options output_var flags)
     list(APPEND compile_options "-fno-unwind-tables")
     list(APPEND compile_options "-fno-asynchronous-unwind-tables")
     list(APPEND compile_options "-fno-rtti")
+    # clang-8+, gcc-12+
+    check_cxx_compiler_flag("-ftrivial-auto-var-init=pattern" LIBC_CC_SUPPORTS_PATTERN_INIT)
```
Comment (lntue) [inline libc/cmake/modules/LLVMLibCObjectRules.cmake:45]: Can you move this check outside of this function, as it is called quite a lot?  Either to the top of the file, or to somewhere in `CheckCompilerFeatures.cmake`.

**Comment by nickdesaulniers** [comment]
```diff
diff --git a/libc/src/__support/str_to_float.h b/libc/src/__support/str_to_float.h
index 8aeb3d2cea03..e8f0cb613b07 100644
--- a/libc/src/__support/str_to_float.h
+++ b/libc/src/__support/str_to_float.h
@@ -30,8 +30,8 @@ namespace LIBC_NAMESPACE {
 namespace internal {
 
 template <class T> struct ExpandedFloat {
-  typename fputil::FPBits<T>::StorageType mantissa;
-  int32_t exponent;
+  typename fputil::FPBits<T>::StorageType mantissa = {};
+  int32_t exponent = 0;
 };
 
 template <class T> struct FloatConvertReturn {
```
knocks out the first difference.

File: libc/cmake/modules/LLVMLibCObjectRules.cmake (hunk 1 above)
Comment (nickdesaulniers) [reply libc/cmake/modules/LLVMLibCObjectRules.cmake]: done in 862dc3c034909e68e971b15af95dfd106233f74e

**Review by lntue — APPROVED** [review]

**Pull request merged**